do_mpc/controller.py

```diff
@@ -30,6 +30,666 @@
+        """Computes discrete gain. 
+        
+        This method computes both finite discrete gain and infinite discrete gain depending on the availability 
+        of prediction horizon. The gain computed using explicit solution for both finite time and infinite time.
+        
+        For finite time:
+            
+            .. math::
+                \\pi(N) &= P_f\\\\
+                K(k) & = -(B'\\pi(k+1)B)^{-1}B'\\pi(k+1)A\\\\
+                \\pi(k) & = Q+A'\\pi(k+1)A-A'\\pi(k+1)B(B'\\pi(k+1)B+R)^{-1}B'\\pi(k+1)A
+       
+        For infinite time:
+            
+            .. math::
+                K & = -(B'PB+P)^{-1}B'PA\\\\
+                P & = Q+A'PA-A'PB(R+B'PB)^{-1}B'PA\\\\
+        
+        For example:
+            
+            ::
+                
+                K = lqr.discrete_gain(A,B)
+                                  
+        :param A: State matrix - constant matrix with no variables
+        :type A: numpy.ndarray
+
+        :param B: Input matrix - constant matrix with no variables
+        :type B: numpy.ndarray                 
+        
+        :return: Gain matrix :math:`K`
+        :rtype: numpy.ndarray
+
+        """
+        #Verifying the availability of cost matrices
+        assert np.shape(self.Q) == (self.model.n_x,self.model.n_x) and np.shape(self.R) == (self.model.n_u,self.model.n_u) , 'Enter tuning parameter Q and R for the lqr problem using set_objective() function.'
+        assert self.model_type == 'discrete', 'convert the model from continous to discrete using model_type_conversion() function.'
+        
+        #calculating finite horizon gain
+        if self.n_horizon !=0:
+            assert self.P.size != 0, 'Terminal cost is required to calculate gain. Enter the required value using set_objective() function.'
+            temp_p = self.P
+            for k in range(self.n_horizon):
+                 K = -np.linalg.inv(np.transpose(B)@temp_p@B+self.R)@np.transpose(B)@temp_p@A
+                 temp_pi = self.Q+np.transpose(A)@temp_p@A-np.transpose(A)@temp_p@B@np.linalg.inv(np.transpose(B)@temp_p@B+self.R)@np.transpose(B)@temp_p@A
+                 temp_p = temp_pi
+            return K
+        
+        #Calculating infinite horizon gain
+        elif self.n_horizon == 0:
+            pi_discrete = solve_discrete_are(A,B, self.Q, self.R)
+            K = -np.linalg.inv(np.transpose(B)@pi_discrete@B+self.R)@np.transpose(B)@pi_discrete@A
+            return K
+    
+    def input_rate_penalization(self):
+        """Computes lqr gain for the input rate penalization mode.
+        
+        This method modifies the state matrix and input matrix according to the input rate penalization method. Due to this objective function also gets modified.
+        The input rate penalization formulation is given as:
+            
+            .. math::
+                x(k+1) = \\tilde{A} x(k) + \\tilde{B}\\Delta u(k)\\\\
+                
+                \\text{where} 
```

Comment:
Please use ``\\quad`` after ``\\text{where}`` to have a bit of space. Otherwise this is the result
<img width="292" alt="Screenshot 2022-09-14 at 18 50 20" src="https://user-images.githubusercontent.com/32864011/190214919-4c572ec9-b247-4d45-99a0-abdcb0b558ef.png">